Delphi 227 walkthrough for advice seekers

Leandro13Silva13 opened the pull request.

Leandro13Silva13 committed:
DELPHI-227 feat:
*Add new walkThroughEnabled field to the User object and set it to true for the user's creation

Leandro13Silva13 committed:
DELPHI-227 feat:
*Update adviceSeekerOf method on UserServiceMapper to check and update the value of walkThroughEnabled for the advice seekers

Leandro13Silva13 committed:
DELPHI-227 feat:
*Fixing and adding tests for the new walkThroughEnabled variable on the User object

Leandro13Silva13 committed:
DELPHI-227 feat:
*Adding tests for the updated adviceSeekerOf method

Leandro13Silva13 committed:
DELPHI-227 feat:
*mvn spotless:apply

Leandro13Silva13 committed:
DELPHI-227 feat:
*mvn spotless:apply

Leandro13Silva13 committed:
DELPHI-227 feat:
*Add the new table column to the changeSet

Leandro13Silva13 committed:
DELPHI-227 feat:
*Add the new table column to the db test files

Leandro13Silva13 committed:
DELPHI-227 feat:
*Fixed the default value of the new column

patric-dosch-vi commented on src/main/resources/db/changelog/changeset/0046_add_walk_through_column_for_advice_seekers/add-walk-through-column-to-advice-seeker.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `userservice`.`user`
```
when it should be active for all new users should the default then not 1? and you need a second query to set it for all where it's null set to 0? As I remember that was the solution you had in the migration sevice?

Leandro13Silva13 replied:
Yes, you are correct, I thought on changing the default value to be 0 instead of 1 to remove the need to add the insert. But the logic is that the value should be 1 for the newly created users, so I agree with you and I will change the default back to 1 and add an insert for the exsiting ones.

patric-dosch-vi replied:
I just mentioned what you had in the MigrationTool. But as for now you anyway set the true to the walkThroughEnabled in the User to true, it is "doppelt gemoppelt" => done twice ;) anyway.

patric-dosch-vi commented on src/main/java/de/caritas/cob/userservice/api/model/User.java:
```diff
@@ -120,12 +120,16 @@ public class User implements TenantAware, NotificationsAware {
       @Size(max = 255) @NonNull String email,
-      boolean languageFormal) {
+      boolean languageFormal,
+      Boolean walkThroughEnabled) {
```
what is the idea here? all "mandatory" fields go into this constructor? or why do you think this is needed here?
I am really not known the logic here, also not for the walkthrough for consultants - it just looks a bit overdressed here

Leandro13Silva13 replied:
Yes, I checked that, excluding the oldID that can be null, all the "mandatory" fields go into this constructor, so I decided to keep the logic there. As I said on another comment, I think the logic for the consultant creation is better, but refactoring would involve more dev and test effort that we cannot afford with the current budget.

roman-vi replied:
we have 20hrs left. I'd say let's come back to this question when you're done with FE changes.

patric-dosch-vi replied:
First, I would always look at the current implementation. At least for me, that helps when the same thing is implemented in the same way.

Coming up with refactoring and budget now is simply too late.
But it's fine how it is, I just was curious WHY we have that change.

patric-dosch-vi replied:
In short: the implementation is fine.

patric-dosch-vi commented on src/main/java/de/caritas/cob/userservice/api/model/User.java:
```diff
@@ -120,12 +120,16 @@ public class User implements TenantAware, NotificationsAware {
   private String notificationsSettings;
 
+  @Column(name = "walk_through_enabled", columnDefinition = "tinyint", nullable = false)
+  private Boolean walkThroughEnabled = true;
```
you put the logic for having it default wise true here? I would put that into the DB like it's also in the consultant-walk-through.sql and manage it in a similar place like here de.caritas.cob.userservice.api.admin.service.consultant.create.CreateConsultantSaga#buildConsultant

Leandro13Silva13 replied:
I added this here to really make sure that the newly created users will have this value true.
And, unfortunatly, I cannot refactor the AS creation to follow the same logic present on de.caritas.cob.userservice.api.admin.service.consultant.create.CreateConsultantSaga#buildConsultant, because the this AS creation is also used for the Anonymous user, and this refactor would increase dev and text effort also increasing the spending. Since, we have to keep the budget in mind, I decided to not refactor to follow the same logic as the consultant creation.

patric-dosch-vi reviewed (commented):
hey @Leandro13Silva13, just had a quick look and added some comments and some hint's to the already existing consultant implementation that I would follow a bit more.

Leandro13Silva13 committed:
DELPHI-227 feat:
*Update the add-walk-through-column-to-advice-seeker.sql to set the column's default value to 1 and insert 0 on the already existing lines.

Leandro13Silva13 committed:
DELPHI-227 feat:
*Update user/data get endpoint to return an extra element on the UserDataResponseDTO response

arrumm reviewed (approved):
approved with a question

roman-vi commented on src/main/java/de/caritas/cob/userservice/api/model/User.java:
```diff
@@ -120,12 +120,16 @@ public class User implements TenantAware, NotificationsAware {
   private String notificationsSettings;
 
+  @Column(name = "walk_through_enabled", columnDefinition = "tinyint", nullable = false)
+  private Boolean walkThroughEnabled = true;
```
is there a real need to use Object type instead of primitive, why it's deviated from languageFormal?

Leandro13Silva13 replied:
Yes because depending on the type of user this value can be true, false or null

roman-vi commented on src/main/java/de/caritas/cob/userservice/api/adapters/web/controller/UserController.java:
```diff
@@ -473,6 +474,7 @@ private EmailNotificationsDTO getEmailNotifications(User user) {
   @Override
   public ResponseEntity<UserDataResponseDTO> getUserData() {
     UserDataResponseDTO partialUserData;
+    Optional<UserSessionResponseDTO> userSession = Optional.empty();
```
I would rather refactor this method to have specific methods for consultant, admin and user returning prepared ```UserDataResponseDTO``` for each case

patric-dosch-vi replied:
I don't think refactoring is really in our focus/budget at the moment :(

roman-vi commented on src/main/java/de/caritas/cob/userservice/api/service/user/UserService.java:
```diff
@@ -71,8 +71,9 @@ public User createUser(
-      String preferredLanguage) {
-    var user = new User(userId, oldId, username, email, languageFormal);
+      String preferredLanguage,
+      Boolean walkThroughEnabled) {
```
we'd definitely better have a builder there

patric-dosch-vi replied:
yes, that's the implementation I had in mind from the Consultant.
Sadly it's more or less everything implemented different :( But I think this we have hard to save :(

roman-vi commented on src/main/java/de/caritas/cob/userservice/api/adapters/web/mapping/UserDtoMapper.java:
```diff
@@ -50,14 +53,22 @@ public UserDataResponseDTO userDataOf(
       twoFactorAuthDTO.setSecret(otpInfoDTO.getOtpSecret());
     }
 
+    Boolean initialInquirySent =
```
Do you handle null e.g. with omitting field returned in DTO?

Leandro13Silva13 replied:
To follow the same logic as the other fields, this must return null to the frontend if the user is a consultant, for example.

roman-vi reviewed (approved):
approved with questions and suggestion

patric-dosch-vi reviewed (approved):
hey @Leandro13Silva13, for me it's fine.
Not sure about Romans null comment, but for me all looks good.

vi-sudo commented:
Task linked: [DELPHI-227 [DELPHI] Task: Walkthrough for Advice Seekers - dev](https://app.clickup.com/t/6656196/DELPHI-227)
